src/Microsoft.OData.Edm/Csdl/Serialization/EdmModelCsdlSchemaWriter.cs

```diff
@@ -47,227 +48,372 @@ internal EdmModelCsdlSchemaWriter(IEdmModel model, Version edmVersion)
 
         internal abstract void WriteIntegerConstantExpressionElement(IEdmIntegerConstantExpression expression);
+        internal abstract Task WriteIntegerConstantExpressionElementAsync(IEdmIntegerConstantExpression expression);
 
         internal abstract void WritePathExpressionElement(IEdmPathExpression expression);
+        internal abstract Task WritePathExpressionElementAsync(IEdmPathExpression expression);
 
         internal abstract void WriteAnnotationPathExpressionElement(IEdmPathExpression expression);
+        internal abstract Task WriteAnnotationPathExpressionElementAsync(IEdmPathExpression expression);
 
         internal abstract void WritePropertyPathExpressionElement(IEdmPathExpression expression);
+        internal abstract Task WritePropertyPathExpressionElementAsync(IEdmPathExpression expression);
 
         internal abstract void WriteNavigationPropertyPathExpressionElement(IEdmPathExpression expression);
+        internal abstract Task WriteNavigationPropertyPathExpressionElementAsync(IEdmPathExpression expression);
 
         internal abstract void WriteIfExpressionElementHeader(IEdmIfExpression expression);
+        internal abstract Task WriteIfExpressionElementHeaderAsync(IEdmIfExpression expression);
 
         internal abstract void WriteIfExpressionElementEnd(IEdmIfExpression expression);
+        internal abstract Task WriteIfExpressionElementEndAsync(IEdmIfExpression expression);
 
         internal abstract void WriteCollectionExpressionElementHeader(IEdmCollectionExpression expression);
+        internal abstract Task WriteCollectionExpressionElementHeaderAsync(IEdmCollectionExpression expression);
 
         internal abstract void WriteCollectionExpressionElementEnd(IEdmCollectionExpression expression);
+        internal abstract Task WriteCollectionExpressionElementEndAsync(IEdmCollectionExpression expression);
 
         internal abstract void WriteLabeledElementHeader(IEdmLabeledExpression labeledElement);
+        internal abstract Task WriteLabeledElementHeaderAsync(IEdmLabeledExpression labeledElement);
 
         internal abstract void WriteLabeledExpressionReferenceExpression(IEdmLabeledExpressionReferenceExpression labeledExpressionReference);
+        internal abstract Task WriteLabeledExpressionReferenceExpressionAsync(IEdmLabeledExpressionReferenceExpression labeledExpressionReference);
 
         internal abstract void WriteTimeOfDayConstantExpressionElement(IEdmTimeOfDayConstantExpression expression);
+        internal abstract Task WriteTimeOfDayConstantExpressionElementAsync(IEdmTimeOfDayConstantExpression expression);
 
         internal abstract void WriteIsTypeExpressionElementHeader(IEdmIsTypeExpression expression, bool inlineType);
+        internal abstract Task WriteIsTypeExpressionElementHeaderAsync(IEdmIsTypeExpression expression, bool inlineType);
 
         internal virtual void WriteIsOfExpressionType(IEdmIsTypeExpression expression, bool inlineType)

```

Comment:
I wonder why these were left empty instead of made abstract like the rest.

Reply:
@habbes 
Noted this as well. Probably there was a plan to provide default implementation for them.

**However, this can be clean up. What do you think?**

Reply:
The purpose of a default implementation is that the child class should still be able to behave reasonably well without overriding the virtual method. If child classes are always expected to override the virtual methods to function correctly, then they should be abstract.

Reply:
I don't know what the impact of making them abstract would be. Maybe it might open a can of worms that derails the PR. For that reason, this is not a blocker for me.

Reply:
Noted